File: android/app/src/main/java/com/reload/prestreloajuda/data/CooldownStore.kt
package com.reload.prestreloajuda.data

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.reload.prestreloajuda.ui.L
import com.reload.prestreloajuda.ui.Lang
import com.reload.prestreloajuda.ui.Strings
import com.reload.prestreloajuda.ui.langPref
import kotlinx.serialization.json.Json

/**
 * Estado do sistema de Cooldown/Alarme (#33) — porte fiel do CooldownStore.swift do Mac.
 * Carrega o catálogo-semente, guarda os personagens e as marcações do usuário (persistido em
 * SharedPreferences, um único blob JSON), calcula tempo restante e AGENDA os alarmes.
 * A VERDADE é sempre o timestamp absoluto (ms) — o alarme e o contador são derivados.
 *
 * O estado é exposto num Compose `State` (`state`), então qualquer Composable que o leia recompõe
 * ao mutar — espelha o `@Published` do `ObservableObject` do Mac.
 */
class CooldownStore(private val appCtx: Context) {

    val catalog: CooldownCatalog = CooldownCatalog.load(appCtx)

    var state by mutableStateOf(loadState())
        private set

    private val prefs = appCtx.getSharedPreferences("cooldowns", Context.MODE_PRIVATE)
    private val json = Json { ignoreUnknownKeys = true; isLenient = true; encodeDefaults = true }

    init {
        reconcile()
    }

    // MARK: - Persistência

    private fun loadState(): CooldownState {
        val blob = appCtx.getSharedPreferences("cooldowns", Context.MODE_PRIVATE)
            .getString(STATE_KEY, null) ?: return CooldownState()
        return try {
            Json { ignoreUnknownKeys = true; isLenient = true }
                .decodeFromString(CooldownState.serializer(), blob)
        } catch (e: Exception) {
            CooldownState()
        }
    }

    /** Aplica um novo estado: carimba `updatedAt` e persiste (equivalente ao `touch()` do Mac). */
    private fun save(next: CooldownState) {
        state = next.copy(updatedAt = nowMs())
        try {
            prefs.edit().putString(STATE_KEY, json.encodeToString(CooldownState.serializer(), state)).apply()
        } catch (_: Exception) {
        }
    }

    private fun lang(): Lang = langPref(appCtx)

    /** Ao abrir: re-agenda os alarmes ainda no futuro (sobrevive a reinício/reboot). */
    private fun reconcile() {
        for (char in state.characters) {
            for (task in shownBattle(char)) if (isBattleActive(char, task)) scheduleBattle(char, task)
            for (berry in shownBerries) if (state.berry[berryKey(char, berry)] != null) scheduleBerry(char, berry)
        }
    }

    // MARK: - Personagens (cadastro)

    fun addCharacter(name: String) {
        val n = name.trim()
        if (n.isEmpty()) return
        val newChar = GameCharacter(id = "char_" + shortId(), name = n.take(30))
        save(state.copy(characters = state.characters + newChar))
    }

    fun renameCharacter(id: String, name: String) {
        val n = name.trim()
        if (n.isEmpty()) return
        save(state.copy(characters = state.characters.map {
            if (it.id == id) it.copy(name = n.take(30)) else it
        }))
    }

    /** Define (ou limpa, com null) a foto do boneco. `pngBase64` = PNG 128×128 já redimensionado. */
    fun setAvatar(id: String, pngBase64: String?) {
        save(state.copy(characters = state.characters.map {
            if (it.id == id) it.copy(avatar = pngBase64) else it
        }))
    }

    fun removeCharacter(id: String) {
        CooldownNotifications.cancelPrefix(appCtx, "cd.battle.$id.")
        CooldownNotifications.cancelPrefix(appCtx, "cd.berry.$id.")
        save(
            state.copy(
                characters = state.characters.filterNot { it.id == id },
                battle = state.battle.filterKeys { !it.startsWith("$id:") },
                berry = state.berry.filterKeys { !it.startsWith("$id:") },
            )
        )
    }

    // MARK: - Tarefas de batalha

    /** Batalhas mostradas por boneco: as obrigatórias; as opcionais ficam num grupo à parte na UI. */
    fun shownBattle(char: GameCharacter): List<BattleTask> = catalog.battleTasks

    private fun battleKey(char: GameCharacter, id: String) = "${char.id}:$id"
    private fun battleNotifId(char: GameCharacter, id: String) = "cd.battle.${char.id}.$id"

    fun isBattleActive(char: GameCharacter, task: BattleTask, now: Long = nowMs()): Boolean =
        state.battle[battleKey(char, task.id)] != null && battleRemainingMs(char, task, now) > 0

    /** Ms restantes; <= 0 (ou nunca marcado) = pronto. */
    fun battleRemainingMs(char: GameCharacter, task: BattleTask, now: Long = nowMs()): Long {
        val markedAt = state.battle[battleKey(char, task.id)] ?: return 0
        return markedAt + (task.hours * 3_600_000).toLong() - now
    }

    fun battleReady(char: GameCharacter, task: BattleTask, now: Long = nowMs()): Boolean =
        battleRemainingMs(char, task, now) <= 0

    enum class BattlePhase { IDLE, RUNNING, READY }

    /** idle = nunca marcado · running = em cooldown · ready = já marcado e liberou. */
    fun battlePhase(char: GameCharacter, task: BattleTask, now: Long = nowMs()): BattlePhase {
        if (state.battle[battleKey(char, task.id)] == null) return BattlePhase.IDLE
        return if (battleRemainingMs(char, task, now) > 0) BattlePhase.RUNNING else BattlePhase.READY
    }

    fun markBattle(char: GameCharacter, task: BattleTask) {
        save(state.copy(battle = state.battle + (battleKey(char, task.id) to nowMs())))
        scheduleBattle(char, task)
    }

    fun clearBattle(char: GameCharacter, task: BattleTask) {
        CooldownNotifications.cancel(appCtx, battleNotifId(char, task.id))
        save(state.copy(battle = state.battle - battleKey(char, task.id)))
    }

    private fun scheduleBattle(char: GameCharacter, task: BattleTask) {
        val markedAt = state.battle[battleKey(char, task.id)] ?: return
        val lang = lang()
        CooldownNotifications.schedule(
            appCtx,
            battleNotifId(char, task.id),
            markedAt + (task.hours * 3_600_000).toLong(),
            Strings.text(L.CdNotifReady, lang).format(task.name.localized(lang)),
            Strings.text(L.CdNotifCharacter, lang).format(char.name),
        )
    }

    // MARK: - Berries

    /** Berries mostradas: as defaultOn + as que o usuário adicionou (enabledBerry). */
    val shownBerries: List<BerryDef>
        get() = catalog.berries.filter { it.defaultOn || state.enabledBerry.contains(it.id) }

    fun addBerry(id: String) {
        if (state.enabledBerry.contains(id)) return
        save(state.copy(enabledBerry = state.enabledBerry + id))
    }

    fun removeBerry(id: String) {
        for (c in state.characters) CooldownNotifications.cancelPrefix(appCtx, "cd.berry.${c.id}.$id.")
        save(
            state.copy(
                enabledBerry = state.enabledBerry.filterNot { it == id },
                berry = state.berry.filterKeys { !it.contains(":$id:") },
            )
        )
    }

    private fun berryKey(char: GameCharacter, berry: BerryDef, plot: Int = 0) =
        "${char.id}:${berry.id}:$plot"

    private fun berryNotifPrefix(char: GameCharacter, berry: BerryDef, plot: Int = 0) =
        "cd.berry.${char.id}.${berry.id}.$plot."

    enum class BerryPhase { EMPTY, GROWING, READY, WILTED }

    /** Instantâneo do canteiro: fase + tempos (plantar/regar/colher) + progresso de rega. */
    data class BerryStatus(
        val phase: BerryPhase,
        val harvestRemainMs: Long,
        val nextWaterRemainMs: Long?,
        val waterPending: Boolean,
        val waterings: Int,      // regas já confirmadas
        val totalWaters: Int,    // regas do tier no ciclo todo
    )

    /** Fase atual + tempos derivados do `plantedAt` + tier. */
    fun berryStatus(char: GameCharacter, berry: BerryDef, plot: Int = 0, now: Long = nowMs()): BerryStatus {
        val p = state.berry[berryKey(char, berry, plot)]
        val tier = catalog.tier(berry.tier)
        if (p == null || tier == null) {
            return BerryStatus(BerryPhase.EMPTY, 0, null, false, 0, 0)
        }
        val total = tier.waterWindowsHours.size
        val harvestAt = p.plantedAt + (tier.growthHours * 3_600_000).toLong()
        val wiltAt = harvestAt + (tier.wiltHours * 3_600_000).toLong()
        var nextWaterRemain: Long? = null
        var pending = false
        if (p.waterings < total) {
            val limit = (tier.waterWindowsHours[p.waterings] * 3_600_000).toLong()
            val fireAt = p.plantedAt + limit - (tier.waterLeadHours * 3_600_000).toLong()
            nextWaterRemain = fireAt - now
            pending = now >= fireAt          // já entrou na janela de regar
        }
        val phase = if (now < harvestAt) BerryPhase.GROWING
        else if (now < wiltAt) BerryPhase.READY else BerryPhase.WILTED
        return BerryStatus(phase, harvestAt - now, nextWaterRemain, pending, p.waterings, total)
    }

    fun plantBerry(char: GameCharacter, berry: BerryDef, plot: Int = 0) {
        save(state.copy(berry = state.berry + (berryKey(char, berry, plot) to BerryProgress(nowMs(), 0))))
        scheduleBerry(char, berry, plot)
    }

    fun waterBerry(char: GameCharacter, berry: BerryDef, plot: Int = 0) {
        val key = berryKey(char, berry, plot)
        val p = state.berry[key] ?: return
        val tier = catalog.tier(berry.tier) ?: return
        if (p.waterings >= tier.waterWindowsHours.size) return
        val np = p.copy(waterings = p.waterings + 1)
        CooldownNotifications.cancel(appCtx, berryNotifPrefix(char, berry, plot) + "w${np.waterings - 1}")
        save(state.copy(berry = state.berry + (key to np)))
    }

    fun harvestBerry(char: GameCharacter, berry: BerryDef, plot: Int = 0) {
        CooldownNotifications.cancelPrefix(appCtx, berryNotifPrefix(char, berry, plot))
        save(state.copy(berry = state.berry - berryKey(char, berry, plot)))
    }

    private fun scheduleBerry(char: GameCharacter, berry: BerryDef, plot: Int = 0) {
        val p = state.berry[berryKey(char, berry, plot)] ?: return
        val tier = catalog.tier(berry.tier) ?: return
        val prefix = berryNotifPrefix(char, berry, plot)
        val lang = lang()
        val body = Strings.text(L.CdNotifCharacter, lang).format(char.name)
        // lembretes de rega (a partir da próxima ainda não feita)
        for (i in p.waterings until tier.waterWindowsHours.size) {
            val fireAt = p.plantedAt + ((tier.waterWindowsHours[i] - tier.waterLeadHours) * 3_600_000).toLong()
            CooldownNotifications.schedule(
                appCtx, prefix + "w$i", fireAt,
                Strings.text(L.CdNotifWater, lang).format(berry.name.localized(lang)), body
            )
        }
        // pronta pra colher
        val harvestAt = p.plantedAt + (tier.growthHours * 3_600_000).toLong()
        CooldownNotifications.schedule(
            appCtx, prefix + "harvest", harvestAt,
            Strings.text(L.CdNotifBerryReady, lang).format(berry.name.localized(lang)), body
        )
        // aviso de wilt (1h antes de murchar)
        val wiltAt = harvestAt + (tier.wiltHours * 3_600_000).toLong()
        CooldownNotifications.schedule(
            appCtx, prefix + "wilt", wiltAt - 3_600_000,
            Strings.text(L.CdNotifWilt, lang).format(berry.name.localized(lang)), body
        )
    }

    // MARK: - Util

    private fun shortId(): String {
        val hex = "0123456789abcdef"
        return (0 until 8).map { hex.random() }.joinToString("")
    }

    companion object {
        private const val STATE_KEY = "cooldowns.state.v2"
    }
}

File: windows/PrestreloAjuda/Services/CooldownStore.cs
using System.IO;
using System.Text.Json;

namespace PrestreloAjuda.Services;

/// Estado do sistema de Cooldown/Alarme (#33): carrega o catálogo-semente, guarda os personagens e
/// as marcações do usuário (persistido em %LOCALAPPDATA%\PrestreloAjuda\cooldowns.state.json),
/// calcula tempo restante e AGENDA os alarmes. Porte 1:1 de CooldownStore.swift do Mac —
/// a verdade é sempre o timestamp absoluto (ms); alarme e contador são derivados.
public sealed class CooldownStore
{
    public CooldownCatalog Catalog { get; }
    public CooldownState State { get; private set; }
    public CooldownNotifications Notifications { get; }

    private static string Dir => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "PrestreloAjuda");
    private static string FilePath => Path.Combine(Dir, "cooldowns.state.json");

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        PropertyNameCaseInsensitive = true,
        WriteIndented = true,
    };

    public CooldownStore(CooldownNotifications? notifications = null)
    {
        Catalog = CooldownCatalog.Load();
        Notifications = notifications ?? new CooldownNotifications();
        State = LoadState();
        Reconcile();
    }

    // MARK: - Persistência

    private static CooldownState LoadState()
    {
        try
        {
            if (File.Exists(FilePath))
                return JsonSerializer.Deserialize<CooldownState>(File.ReadAllText(FilePath), JsonOpts) ?? new CooldownState();
        }
        catch { /* decoder tolerante: qualquer falha → estado zerado */ }
        return new CooldownState();
    }

    private void Touch()
    {
        State.UpdatedAt = Cd.NowMs();
        try
        {
            Directory.CreateDirectory(Dir);
            File.WriteAllText(FilePath, JsonSerializer.Serialize(State, JsonOpts));
        }
        catch { /* persistência é best-effort */ }
    }

    /// Ao abrir: re-agenda os alarmes ainda no futuro (sobrevive a fechar/reabrir o app).
    private void Reconcile()
    {
        foreach (var ch in State.Characters)
        {
            foreach (var task in ShownBattle(ch))
                if (IsBattleActive(ch, task)) ScheduleBattle(ch, task);
            foreach (var berry in ShownBerries)
                if (State.Berry.ContainsKey(BerryKey(ch, berry))) ScheduleBerry(ch, berry);
        }
    }

    // MARK: - Personagens (cadastro)

    public void AddCharacter(string name)
    {
        var n = name.Trim();
        if (n.Length == 0) return;
        State.Characters.Add(new GameCharacter { Id = "char_" + ShortId(), Name = Clip(n, 30) });
        Touch();
    }

    public void RenameCharacter(string id, string name)
    {
        var n = name.Trim();
        if (n.Length == 0) return;
        var c = State.Characters.FirstOrDefault(x => x.Id == id);
        if (c == null) return;
        c.Name = Clip(n, 30);
        Touch();
    }

    /// Define (ou limpa, com null) a foto do boneco. `pngBase64` = PNG 128×128 já redimensionado.
    public void SetAvatar(string id, string? pngBase64)
    {
        var c = State.Characters.FirstOrDefault(x => x.Id == id);
        if (c == null) return;
        c.Avatar = pngBase64;
        Touch();
    }

    public void RemoveCharacter(string id)
    {
        State.Characters.RemoveAll(x => x.Id == id);
        foreach (var k in State.Battle.Keys.Where(k => k.StartsWith(id + ":", StringComparison.Ordinal)).ToList())
            State.Battle.Remove(k);
        foreach (var k in State.Berry.Keys.Where(k => k.StartsWith(id + ":", StringComparison.Ordinal)).ToList())
            State.Berry.Remove(k);
        Notifications.CancelPrefix($"cd.battle.{id}.");
        Notifications.CancelPrefix($"cd.berry.{id}.");
        Touch();
    }

    // MARK: - Tarefas de batalha

    /// Batalhas mostradas por boneco: as obrigatórias (catalog.battleTasks). As opcionais ficam num
    /// grupo à parte na UI (via Catalog.OptionalTasks).
    public IReadOnlyList<BattleTask> ShownBattle(GameCharacter ch) => Catalog.BattleTasks;

    private static string BattleKey(GameCharacter ch, string id) => $"{ch.Id}:{id}";
    private static string BattleNotifId(GameCharacter ch, string id) => $"cd.battle.{ch.Id}.{id}";

    public bool IsBattleActive(GameCharacter ch, BattleTask task)
        => State.Battle.ContainsKey(BattleKey(ch, task.Id)) && BattleRemainingMs(ch, task) > 0;

    /// Ms restantes; <= 0 (ou nunca marcado) = pronto.
    public double BattleRemainingMs(GameCharacter ch, BattleTask task)
    {
        if (!State.Battle.TryGetValue(BattleKey(ch, task.Id), out var markedAt)) return 0;
        return markedAt + task.Hours * 3_600_000 - Cd.NowMs();
    }

    public bool BattleReady(GameCharacter ch, BattleTask task) => BattleRemainingMs(ch, task) <= 0;

    public enum BattlePhase { Idle, Running, Ready }

    /// Idle = nunca marcado · Running = em cooldown · Ready = já marcado e liberou.
    public BattlePhase Phase(GameCharacter ch, BattleTask task)
    {
        if (!State.Battle.ContainsKey(BattleKey(ch, task.Id))) return BattlePhase.Idle;
        return BattleRemainingMs(ch, task) > 0 ? BattlePhase.Running : BattlePhase.Ready;
    }

    public void MarkBattle(GameCharacter ch, BattleTask task)
    {
        State.Battle[BattleKey(ch, task.Id)] = Cd.NowMs();
        Touch();
        ScheduleBattle(ch, task);
    }

    public void ClearBattle(GameCharacter ch, BattleTask task)
    {
        State.Battle.Remove(BattleKey(ch, task.Id));
        Notifications.Cancel(BattleNotifId(ch, task.Id));
        Touch();
    }

    private void ScheduleBattle(GameCharacter ch, BattleTask task)
    {
        if (!State.Battle.TryGetValue(BattleKey(ch, task.Id), out var markedAt)) return;
        Notifications.Schedule(
            BattleNotifId(ch, task.Id),
            markedAt + task.Hours * 3_600_000,
            string.Format(Strings.T(L.CdNotifReady), task.Name.Localized),
            string.Format(Strings.T(L.CdNotifCharacter), ch.Name));
    }

    // MARK: - Berries

    /// Berries mostradas: as defaultOn + as que o usuário adicionou (EnabledBerry).
    public IReadOnlyList<BerryDef> ShownBerries =>
        Catalog.Berries.Where(b => b.DefaultOn || State.EnabledBerry.Contains(b.Id)).ToList();

    public void AddBerry(string id)
    {
        if (State.EnabledBerry.Contains(id)) return;
        State.EnabledBerry.Add(id);
        Touch();
    }

    public void RemoveBerry(GameCharacter? ch, string id)
    {
        State.EnabledBerry.RemoveAll(x => x == id);
        // limpa plantios dessa berry (de todos os bonecos) + cancela alarmes
        foreach (var k in State.Berry.Keys.Where(k => k.Contains($":{id}:", StringComparison.Ordinal)).ToList())
            State.Berry.Remove(k);
        foreach (var c in State.Characters) Notifications.CancelPrefix($"cd.berry.{c.Id}.{id}.");
        Touch();
    }

    private static string BerryKey(GameCharacter ch, BerryDef berry, int plot = 0) => $"{ch.Id}:{berry.Id}:{plot}";
    private static string BerryNotifPrefix(GameCharacter ch, BerryDef berry, int plot = 0) => $"cd.berry.{ch.Id}.{berry.Id}.{plot}.";

    public enum BerryPhase { Empty, Growing, Ready, Wilted }

    /// Instantâneo do canteiro: fase + tempos (plantar/regar/colher) + progresso de rega.
    public sealed class BerryStatus
    {
        public BerryPhase Phase;
        public double HarvestRemainMs;
        public double? NextWaterRemainMs;
        public bool WaterPending;
        public int Waterings;      // regas já confirmadas
        public int TotalWaters;    // regas do tier no ciclo todo
    }

    /// Fase atual + tempos derivados do PlantedAt + tier. (Nome BerryStat p/ não colidir com o tipo.)
    public BerryStatus BerryStat(GameCharacter ch, BerryDef berry, int plot = 0)
    {
        if (!State.Berry.TryGetValue(BerryKey(ch, berry, plot), out var p) ||
            Catalog.FindTier(berry.Tier) is not { } tier)
        {
            return new BerryStatus { Phase = BerryPhase.Empty };
        }
        double now = Cd.NowMs();
        int total = tier.WaterWindowsHours.Count;
        double harvestAt = p.PlantedAt + tier.GrowthHours * 3_600_000;
        double wiltAt = harvestAt + tier.WiltHours * 3_600_000;
        // próxima rega ainda não confirmada
        double? nextWaterRemain = null;
        bool pending = false;
        if (p.Waterings < total)
        {
            double limit = tier.WaterWindowsHours[p.Waterings] * 3_600_000;
            double fireAt = p.PlantedAt + limit - tier.WaterLeadHours * 3_600_000;
            nextWaterRemain = fireAt - now;
            pending = now >= fireAt;   // já entrou na janela de regar
        }
        var phase = now < harvestAt ? BerryPhase.Growing : (now < wiltAt ? BerryPhase.Ready : BerryPhase.Wilted);
        return new BerryStatus
        {
            Phase = phase,
            HarvestRemainMs = harvestAt - now,
            NextWaterRemainMs = nextWaterRemain,
            WaterPending = pending,
            Waterings = p.Waterings,
            TotalWaters = total,
        };
    }

    public void PlantBerry(GameCharacter ch, BerryDef berry, int plot = 0)
    {
        State.Berry[BerryKey(ch, berry, plot)] = new BerryProgress { PlantedAt = Cd.NowMs(), Waterings = 0 };
        Touch();
        ScheduleBerry(ch, berry, plot);
    }

    public void WaterBerry(GameCharacter ch, BerryDef berry, int plot = 0)
    {
        var key = BerryKey(ch, berry, plot);
        if (!State.Berry.TryGetValue(key, out var p) || Catalog.FindTier(berry.Tier) is not { } tier
            || p.Waterings >= tier.WaterWindowsHours.Count) return;
        p.Waterings += 1;
        State.Berry[key] = p;
        Notifications.Cancel(BerryNotifPrefix(ch, berry, plot) + "w" + (p.Waterings - 1));
        Touch();
    }

    public void HarvestBerry(GameCharacter ch, BerryDef berry, int plot = 0)
    {
        State.Berry.Remove(BerryKey(ch, berry, plot));
        Notifications.CancelPrefix(BerryNotifPrefix(ch, berry, plot));
        Touch();
    }

    private void ScheduleBerry(GameCharacter ch, BerryDef berry, int plot = 0)
    {
        if (!State.Berry.TryGetValue(BerryKey(ch, berry, plot), out var p) ||
            Catalog.FindTier(berry.Tier) is not { } tier) return;
        var prefix = BerryNotifPrefix(ch, berry, plot);
        var body = string.Format(Strings.T(L.CdNotifCharacter), ch.Name);
        // lembretes de rega (a partir da próxima ainda não feita)
        for (int i = p.Waterings; i < tier.WaterWindowsHours.Count; i++)
        {
            double fireAt = p.PlantedAt + (tier.WaterWindowsHours[i] - tier.WaterLeadHours) * 3_600_000;
            Notifications.Schedule(prefix + "w" + i, fireAt,
                string.Format(Strings.T(L.CdNotifWater), berry.Name.Localized), body);
        }
        // pronta pra colher
        double harvestAt = p.PlantedAt + tier.GrowthHours * 3_600_000;
        Notifications.Schedule(prefix + "harvest", harvestAt,
            string.Format(Strings.T(L.CdNotifBerryReady), berry.Name.Localized), body);
        // aviso de wilt (1h antes de murchar)
        double wiltAt = harvestAt + tier.WiltHours * 3_600_000;
        Notifications.Schedule(prefix + "wilt", wiltAt - 3_600_000,
            string.Format(Strings.T(L.CdNotifWilt), berry.Name.Localized), body);
    }

    // MARK: - Util

    private static string Clip(string s, int max) => s.Length > max ? s.Substring(0, max) : s;

    private static readonly Random Rng = new();
    private static string ShortId()
    {
        const string hex = "0123456789abcdef";
        var chars = new char[8];
        for (int i = 0; i < chars.Length; i++) chars[i] = hex[Rng.Next(hex.Length)];
        return new string(chars);
    }
}
